src/D2L.CodeStyle.Analyzers/ApiUsage/RpcAnalyzer.cs

```diff
@@ -26,171 +28,205 @@ internal sealed class RpcAnalyzer : DiagnosticAnalyzer {
-				rpcPostContextType.Equals( firstParamType, SymbolEqualityComparer.Default ) ||
-				rpcPostContextBaseType.Equals( firstParamType, SymbolEqualityComparer.Default );
-
-			if( !firstParamIsReasonableType ) {
+			if( !rpcTypes.RpcContexts.Contains( rpcContext.Symbol.Type ) ) {
 				context.ReportDiagnostic(
-					Diagnostic.Create( Diagnostics.RpcContextFirstArgument, firstParam.GetLocation() )
+					Diagnostic.Create( Diagnostics.RpcContextFirstArgument, rpcContext.Syntax.GetLocation() )
 				);
-			} else if( dependencyAttributeType != null
-				&& IsMarkedDependency( dependencyAttributeType, firstParam, context.SemanticModel )
-			) {
+				return;
+			}
+
+			if( IsMarkedAsDependency( rpcContext.Symbol, rpcTypes ) ) {
 				context.ReportDiagnostic(
-					Diagnostic.Create( Diagnostics.RpcContextMarkedDependency, firstParam.GetLocation() )
+					Diagnostic.Create( Diagnostics.RpcContextMarkedDependency, rpcContext.Syntax.GetLocation() )
 				);
 			}
 		}
 
-		private static void CheckThatDependencyArgumentsAreSortedCorrectly(
+		private static void CheckThatRpcParametersAreValid(
+			SyntaxNodeAnalysisContext context,
+			ImmutableArray<(IParameterSymbol Symbol, ParameterSyntax Syntax)> parameters,
+			RpcTypes rpcTypes
+		) {
+			foreach( var parameter in parameters ) {
+				bool isValidParameterType = IsValidParameterType( context, parameter.Symbol.Type, rpcTypes );
+				if( !isValidParameterType ) {
+					context.ReportDiagnostic(
+						Diagnostic.Create( Diagnostics.RpcInvalidParameterType, parameter.Syntax.Type.GetLocation() )
+					);
+				}
+			}
+		}
+
+		private static ((IParameterSymbol Symbol, ParameterSyntax Syntax) rpcContext, ImmutableArray<(IParameterSymbol Symbol, ParameterSyntax Syntax)> dependencies, ImmutableArray<(IParameterSymbol Symbol, ParameterSyntax Syntax)> parameters) Split(
 			SyntaxNodeAnalysisContext context,
-			SeparatedSyntaxList<ParameterSyntax> ps,
-			INamedTypeSymbol dependencyAttributeType
+			MethodDeclarationSyntax method,
+			RpcTypes rpcTypes
 		) {
-			bool doneDependencies = false;
-			foreach( var param in ps.Skip( 1 ) ) {
-				var isDep = IsMarkedDependency( dependencyAttributeType, param, context.SemanticModel );
-
-				if( !isDep && !doneDependencies ) {
-					doneDependencies = true;
-				} else if( isDep && doneDependencies ) {
-					context.ReportDiagnostic( Diagnostic.Create( Diagnostics.RpcArgumentSortOrder, param.GetLocation() ) );
+			var parameters = method.ParameterList.Parameters;
+
+			if( parameters.Count == 0 ) {
+				return (default, ImmutableArray<(IParameterSymbol, ParameterSyntax)>.Empty, ImmutableArray<(IParameterSymbol, ParameterSyntax)>.Empty);
+			}
+
+			SemanticModel model = context.SemanticModel;
+
+			int index = 0;
+
+			ParameterSyntax rpcContextSyntax = parameters[ index++ ];
+			IParameterSymbol rpcContext = model.GetDeclaredSymbol( rpcContextSyntax );
+			var dependenciesBuilder = ImmutableArray.CreateBuilder<(IParameterSymbol, ParameterSyntax)>();

```

Comment:
Don't actually use the dependencies. Could not collect them.

I'll change that to shouldn't as to not waste analyzer cycles. Activation Diagnostics would make sure the type was actually registered. We do that elsewhere now.